### ISSUE TITLE
[B4.1] Create Module trait in shared/core/module.mojo for layer composition

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2305]
## Overview
Create a Module trait that provides a standard interface for neural network layers and models.

## Location
- Create new `shared/core/module.mojo`

## Purpose
The Module trait enables:
- Consistent interface for all layers (Linear, Conv2D, etc.)
- Easy parameter collection for optimizers
- Nested module support (models containing layers)
- Training vs eval mode switching

## Deliverables
- [ ] Create `shared/core/module.mojo` with Module trait:
  - forward() - forward pass computation
  - parameters() - return list of trainable parameters
  - train() / eval() - switch training mode
  - children() - return nested modules (optional)
- [ ] Implement ModuleList for containing multiple modules
- [ ] Add examples showing trait usage
- [ ] Export from `shared/core/__init__.mojo`

## Interface Design
```mojo
trait Module:
    fn forward(mut self, input: ExTensor) raises -> ExTensor
    fn parameters(self) -> List[Variable]
    fn train(mut self)
    fn eval(mut self)
```

## Success Criteria
- Module trait works with existing layers
- Easy to implement for new layers
- Supports model composition
- CI passes

## Dependencies
- None

## Track
Track B: Missing Features (B4.1 of 8)
